### ISSUE TITLE
Build fails on impossible Typescript error with RxJS

### ISSUE THREAD

**Issue by mgporter** [angular/angular-cli#28079]
### Command

build, serve

### Is this a regression?

- [ ] Yes, this behavior used to work in the previous version

### Description

The Angular cli refuses to build a project with error "TS2345: Argument of type ... is not assignable to parameter of type ...", even though Typescript itself is able to correctly deduce that no type issue is occurring. This issue occurs with RxJS.

### Minimal Reproduction

In the code below, null values for `id` are filtered out of the pipe, so that the type of `id` inside the map function is only `number`. Typescript correctly produces no error (at least for me, using vs code). Nonetheless, the Angular build still fails, thinking that `id` is still of a nullable type.

```
  myFunction(): void {
    const objectWithNullableProperty: {id: number | null } = {id: 243};

    of(objectWithNullableProperty).pipe(
      map(obj => obj.id),
      filter(id => id != null),
      map(id => this.doSomethingWithId(id))
    ).subscribe();
  }

  doSomethingWithId(id: number): void {
    console.log(id);
  }
```

### Exception or Error

```text
Build fails with error:

TS2345: Argument of type 'number | null' is not assignable to parameter of type 'number'.
  Type 'null' is not assignable to type 'number'.
```

### Your Environment

```text
Angular CLI: 18.1.0
Node: 20.12.1
Package Manager: npm 10.7.0
OS: win32 x64

Angular: 18.1.0
... animations, cli, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1801.0
@angular-devkit/build-angular   18.1.0
@angular-devkit/core            18.1.0
@angular-devkit/schematics      18.1.0
@schematics/angular             18.1.0
rxjs                            7.8.1
typescript                      5.4.5
zone.js                         0.14.7
```

**Comment by JeanMeche** [comment]
This is a feature that landed with TS 5.5. You need to update your dependency.

**Comment by mgporter** [comment]
Upgrading to typescript 5.5 works, thank you!! That threw me off because there isn't an error in the IDE; it's only an error on build.
